---
title: "Correlation analysis"
author: "Fernanda Terrazas"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
### Correlation between Hydrogen and Methane
```{r}
######## DO CH4-H2 CORRELATION WITH LOG VALUES
plot(df_log_breath[6:7])

# library(psych)
pairs.panels(df_log_breath[6:7])

shapiro.test(df_log_breath$H2) # not normal
shapiro.test(df_log_breath$CH4) # not normal

# Test significance of correlation - nonparametric
corr_test_sp <- cor.test(df_log_breath$H2, df_log_breath$CH4, method ="spearman")
corr_test_sp
corr_test_kn <- cor.test(df_log_breath$H2, df_log_breath$CH4, method ="kendall")
corr_test_kn

# A positive rho value indicates that as one variable increases, the other tends to increase as well. The value of rho ranges between # -1 and 1. Closer to -1 means a strong negative correlation, closer to 1 means a strong positive correlation, and closer to 0 means # a weaker correlation.

# The analysis indicates a statistically significant, moderate, positive correlation between CH4 and H2. Keep in mind that correlation does not imply causation.

################### RUN CORRELATION WITHOUT OUTLIERS -- use 'outliers'
df_log_breath_no.outliers <- df_log_breath %>%
  filter(!Subject %in% outliers$Subject)

corr_test_sp_no.outliers <- cor.test(df_log_breath_no.outliers$H2, df_log_breath_no.outliers$CH4, method ="spearman")
corr_test_sp_no.outliers

```
### **DO BREATH GAS-MICROBIAL ABUNDANCES CORRELATIONS WITH AUC VALUES**

Change to hydro_meth_df to avoid major changes to code; also, change AUC columns names to match H2 and CH4.
```{r}
hydro_meth_df <- df_h2ch4_auc
head(hydro_meth_df)

# Rename columns
colnames(hydro_meth_df) <- c('Subject','Treatment','Day','H2','CH4','SampleID')
head(hydro_meth_df)

# Reorder columns
order <- c('SampleID','Subject','Treatment','Day','H2','CH4')
hydro_meth_df <- hydro_meth_df %>%
   select(all_of(order))
head(hydro_meth_df)

# Count number of NAs?
na_count <- sum(is.na(hydro_meth_df$SampleID))
print(na_count)

```
## H2 & CH4 Outliers
```{r}
outliers <- subset(hydro_meth_df, Subject %in% c(950, 962, 921, 735)) # 737, 989?

library(ggrepel)

p_h2_ch4_relat <- ggplot(hydro_meth_df, aes(x = H2, y = CH4, color = Treatment, shape = Treatment)) +
  geom_point(size = 3) +
  geom_text_repel(data = outliers, aes(x = H2, y = CH4, label = Subject), nudge_x = 0.1, nudge_y = 0.1, color = "black", size = 3) +
  xlab("Hydrogen (ppm)") +
  ylab("Methane (ppm)") +
  theme_classic()
p_h2_ch4_relat

```
Save plot
```{r}
ggsave(filename = "~/Documents/Programming/Microbiome/microbiome-ocp/data-analyses/results/h2_ch4_corr/p_h2_ch4_relationship.png", plot = p_h2_ch4_relat, width = 5, height = 4, dpi = 300)

```

Compare effects of Treatment (OCs) on H2 and CH4 levels -- **(exclude outliers?)**

```{r}
# Filter out outliers??????
hydro_meth_treat_no.outliers <- hydro_meth_df[!(hydro_meth_df$SampleID %in% outliers$SampleID),]
hydro_meth_treat_no.outliers <- hydro_meth_df[(hydro_meth_df$SampleID %in% group_control_oc$SampleID),]
hydro_meth_treat_no.outliers <- hydro_meth_df

# Renaming the columns from H2 to Hydrogen, CH4 to Methane for aesthetics
hydro_meth_treat_no.outliers <- hydro_meth_treat_no.outliers %>%
  rename(
    Methane = CH4,
    Hydrogen = H2
  )

h2_oc.use_p <- ggplot(hydro_meth_treat_no.outliers, aes(x = Treatment, y = Hydrogen)) +
  geom_boxplot(aes (fill = Treatment), colour = "black") +
  scale_fill_manual(values = c("#332288","#E066A6","#E6A316", "#256769")) +
  ylab("Hydrogen (ppm)") +
  theme_pubclean()
h2_oc.use_p

ch4_oc.use_p <- ggplot(hydro_meth_treat_no.outliers, aes(x = Treatment, y = Methane)) +
  geom_boxplot(aes (fill = Treatment), colour = "black") +
  scale_fill_manual(values = c("#332288","#E066A6","#E6A316", "#256769")) +
  ylab("Methane (ppm)") +
  theme_pubclean()
ch4_oc.use_p

# Reshape the data into a long format
hydro_meth_treat_no.outliers_long <- hydro_meth_treat_no.outliers %>%
  pivot_longer(cols = c(Hydrogen, Methane), names_to = "Gas", values_to = "Value")

h2_ch4_control_oc_p <- ggplot(hydro_meth_treat_no.outliers_long, aes(x = Treatment, y = Value, fill = Treatment)) +
  geom_boxplot(colour = "black") +
  scale_fill_manual(values = c("User" = "#E066A6", "Control" = "#E6A316")) + # Change according to Diet levels
  facet_wrap(~ Gas, scales = "free") +
  ylab("Concentration (ppm)") +
  xlab("") +
  theme_pubclean()
h2_ch4_control_oc_p

###############################################
##### MODIFIED ABOVE TO INCLUDE DAY INTERACTION
###############################################

h2_ch4_control_oc_p <- ggplot(hydro_meth_treat_no.outliers_long, 
                              aes(x = Treatment, y = Value, fill = interaction(Treatment, Day))) +
  geom_boxplot(position = "dodge", colour = "black") +
  scale_fill_manual(
    values = c("#E6A316", "#E066A6", "#e8d1a0", "#e0bccf"), 
    name = "Treatment:Day",
    labels = c("Day 1 Control", "Day 1 User", "Day 21 Control", "Day 21 User") # Modify as needed
  ) +
  facet_wrap(~ Gas, scales = "free") +
  ylab("Concentration (ppm)") +
  xlab("") +
  # theme(axis.text.x = element_text(angle = 45, hjust = 1)) # Optionally rotate x-axis labels for better readability
  theme_pubclean()
h2_ch4_control_oc_p


# h2_ch4_diet_treat_p <- egg::ggarrange(h2_ch4_control_oc_p, h2_ch4_prune_no_prune_p, labels = c("A", "B"), 
#                          ncol = 2, nrow = 1, heights = c(3))
# h2_ch4_diet_treat_p

```
Save plot
```{r}
ggsave(filename = "~/Documents/Programming/Microbiome/microbiome-ocp/data-analyses/results/h2_ch4_corr/p_h2_ch4_levels_treat.png", plot = h2_ch4_control_oc_p, width = 5, height = 4, dpi = 300)

# # combined plot
# ggsave(filename = "~/Documents/Programming/Microbiome/microbiome-ocp/data-analyses/results/h2_ch4_corr/p_h2_ch4_levels_treat_diet.png", plot = h2_ch4_diet_treat_p, width = 8, height = 4, dpi = 300)

```
## Get H2/CH4 ratios
```{r}
# subset data according to group -- use hydro_meth_diet_no.outliers?
# hydro_meth_pnp <- hydro_meth_df[row.names(hydro_meth_df) %in% map$SampleID, ]
head(hydro_meth_df)

# save original df as a new variable
hydro_meth_df_orig <- hydro_meth_df

# calculate the H2 to CH4 ratio
hydro_meth_df$CH4_to_H2 <- hydro_meth_df$CH4/hydro_meth_df$H2
head(hydro_meth_df)

```
## Subset data
```{r}
# subset clr-transformed data according to 'hydro_meth_df'
species_tx_h2_ch4 <- species_tx[, colnames(species_tx) %in% hydro_meth_df$SampleID]
head(species_tx_h2_ch4)

```
## Get H2/CH4 correlations with clr species abundances
```{r}
# Transpose the microbial abundances dataframe and convert SampleID column names to a separate column
species_tx_h2_ch4_t <- data.frame(t(species_tx_h2_ch4))
head(species_tx_h2_ch4_t)

# Convert row names to a column
species_tx_h2_ch4_t_rn <- rownames_to_column(species_tx_h2_ch4_t, var = "SampleID")
head(species_tx_h2_ch4_t_rn)

# merge transposed abundances with map data
hydro_meth_species_tx <- merge(hydro_meth_df, species_tx_h2_ch4_t_rn, by = "SampleID")
head(hydro_meth_species_tx)

```
Compute correlations -- **CH4_to_H2**
```{r}
# compute correlations between ch4/h2 ratio and species abundances
# ch4:h2 correlations
corr_ch4_to_h2 <- sapply(hydro_meth_species_tx[,9:ncol(hydro_meth_species_tx)], function(species) {
  test_result <- cor.test(hydro_meth_species_tx$CH4_to_H2, species, method = 'spearman') # method = 'pearson'
  c(Estimate = test_result$estimate, Pvalue = test_result$p.value)
}, simplify = "data.frame")

# h2 correlations
corr_h2 <- sapply(hydro_meth_species_tx[,9:ncol(hydro_meth_species_tx)], function(species) {
  test_result <- cor.test(hydro_meth_species_tx$H2, species, method = 'spearman') # method = 'pearson'
  c(Estimate = test_result$estimate, Pvalue = test_result$p.value)
}, simplify = "data.frame")

# ch4 correlations
corr_ch4 <- sapply(hydro_meth_species_tx[,9:ncol(hydro_meth_species_tx)], function(species) {
  test_result <- cor.test(hydro_meth_species_tx$CH4, species, method = 'spearman') # method = 'pearson'
  c(Estimate = test_result$estimate, Pvalue = test_result$p.value)
}, simplify = "data.frame")



# Convert results list to dataframe
# ch4:h2 correlations dataframe
corr_results_ch4_to_h2 <- as.data.frame(corr_ch4_to_h2)
corr_results_ch4_to_h2 <- data.frame(t(corr_results_ch4_to_h2))
colnames(corr_results_ch4_to_h2) <- c("Correlation", "Pvalue")

# h2 correlations dataframe
corr_results_h2 <- as.data.frame(corr_h2)
corr_results_h2 <- data.frame(t(corr_results_h2))
colnames(corr_results_h2) <- c("Correlation", "Pvalue")

# ch4 correlations dataframe
corr_results_ch4 <- as.data.frame(corr_ch4)
corr_results_ch4 <- data.frame(t(corr_results_ch4))
colnames(corr_results_ch4) <- c("Correlation", "Pvalue")



# Add species names (rownames) as a new column
# ch4:h2 correlations + species names
corr_results_ch4_to_h2 <- rownames_to_column(corr_results_ch4_to_h2, var = "Species")
# rownames(corr_results_ch4_to_h2) <- NULL
head(corr_results_ch4_to_h2)

# h2 correlations + species names
corr_results_h2 <- rownames_to_column(corr_results_h2, var = "Species")
# rownames(corr_results_h2) <- NULL
head(corr_results_h2)

# ch4 correlations + species names
corr_results_ch4 <- rownames_to_column(corr_results_ch4, var = "Species")
# rownames(corr_results_ch4) <- NULL
head(corr_results_ch4)

```
Save microbial abundance correlations as .csv file
```{r}
write.csv(corr_results_ch4_to_h2, "~/Documents/Programming/Microbiome/microbiome-ocp/data-analyses/results/h2_ch4_corr/corr_results_ch4_to_h2.csv", row.names = FALSE)
write.csv(corr_results_h2, "~/Documents/Programming/Microbiome/microbiome-ocp/data-analyses/results/h2_ch4_corr/corr_results_h2.csv", row.names = FALSE)
write.csv(corr_results_ch4, "~/Documents/Programming/Microbiome/microbiome-ocp/data-analyses/results/h2_ch4_corr/corr_results_ch4.csv", row.names = FALSE)

```
Rename taxa to simply Species name
```{r}
# Define a function to find the most specific taxon
find_most_specific_taxon <- function(taxon_string) {
  # Split the string into individual taxon
  split_taxon <- strsplit(taxon_string, "\\.")[[1]]

  # Remove the level identifiers
  named_taxon <- gsub("^.*__ ", "", split_taxon)

  # Reverse the order
  rev_taxon <- rev(named_taxon)

  # Find the first taxon that doesn't say 'uncultured' or 'metagenome'
  specific_taxon <- rev_taxon[!rev_taxon %in% c('uncultured',
                                                'unidentified',
                                                'uncultured_bacterium',
                                                'metagenome',
                                                'uncultured_organism',
                                                'human_gut',
                                                'gut_metagenome',
                                                '_bacterium')][1]

  return(specific_taxon)
}

# corr_results_ch4_to_h2
# corr_results_h2
# corr_results_ch4

species_c4m <- corr_results_ch4_to_h2
species_c4m$Species <- corr_results_ch4_to_h2$Species

# View the updated data frame
head(species_c4m)

# Remove extra "__." 
species_c4m$Species <- gsub("\\.__$", "", species_c4m$Species)
species_c4m$Species <- gsub("\\.__$", "", species_c4m$Species)
species_c4m$Species <- gsub("\\.__$", "", species_c4m$Species)
species_c4m$Species <- gsub("\\.__$", "", species_c4m$Species)
species_c4m$Species <- gsub("\\.__$", "", species_c4m$Species)

# Remove "." where it's not followed by any more characters
species_c4m$Species <- gsub("\\.$", "", species_c4m$Species)

# Remove "." where followed by a single "_"
species_c4m$Species <- gsub("\\._(?!_)", "_", species_c4m$Species, perl=TRUE)

# Remove "." when preceded by two "__"
species_c4m$Species <- gsub("(?<=__)\\.", "", species_c4m$Species, perl=TRUE)

# Add a space after each of the specified patterns
species_c4m$Species <- gsub("(d__|p__|c__|o__|f__|g__|s__)", "\\1 ", species_c4m$Species)

# Remove "." from species/genera like UCG.003 so it reads UCG003
species_c4m$Species <- gsub("([a-zA-Z])\\.([0-9])", "\\1\\2", species_c4m$Species)

# Check
head(species_c4m$Species)

# Apply function to Species count table
species_c4m$Species <- sapply(species_c4m$Species, find_most_specific_taxon)

# Clean up special characters left in names
species_c4m$Species <- gsub("\\[|\\]", "", species_c4m$Species)

# Replace underscores with spaces
species_c4m$Species <- gsub("_", " ", species_c4m$Species)

# Replace "." in sp.
species_c4m$Species <- gsub("sp$", "sp.", species_c4m$Species)



#####
head(corr_results_ch4_to_h2)
corr_results_ch4_to_h2 <- species_c4m
head(corr_results_ch4_to_h2)

```
Correlation log plot -- **CH4_to_H2**
```{r}
library(ggrepel)

ggplot(corr_results_ch4_to_h2, aes(x=Correlation, y=-log10(Pvalue))) + geom_point() + labs(x="Correlation", y="-log10(Pvalue)") + geom_hline(yintercept=-log10(0.05), linetype="dashed", color = "red")
                
# Create a new column in the dataframe to indicate significance
corr_results_ch4_to_h2$Significant <- ifelse(corr_results_ch4_to_h2$Pvalue < 0.05, "Significant", "Not significant")

# Create the volcano plot
p_ch4_to_h2_corr <- ggplot(corr_results_ch4_to_h2, aes(x=Correlation, y=-log10(Pvalue), color=Significant)) + 
  geom_point() + 
  geom_hline(yintercept=-log10(0.05), linetype="dashed", color = "red") +
  labs(x="Spearman ρ", y="-log10(p-value)", color="Significance") + 
  theme_bw() + 
  geom_text_repel(data = subset(corr_results_ch4_to_h2, Pvalue < 0.05), 
                  aes(label=Species), 
                  size=3, 
                  box.padding = unit(0.35, "lines"), 
                  max.overlaps = 10, 
                  color = "black",
                  fontface = "bold",
                  segment.color = "grey",
                  segment.alpha = 0.7) +
  ggtitle("") + # CH4/H2 ratio
  theme_pubclean() +
  theme(legend.position = "none")
p_ch4_to_h2_corr

# # Correct 'Species' names if needed
# corr_results_ch4_to_h2[154, "Species"] <- "Ruminococcaceae"

```
Correlation log plot -- **H2**
```{r}
# Create a new column in the dataframe to indicate significance
corr_results_h2$Significant <- ifelse(corr_results_h2$Pvalue < 0.05, "Significant", "Not significant")

# Create the volcano plot
p_h2_corr <- ggplot(corr_results_h2, aes(x=Correlation, y=-log10(Pvalue), color=Significant)) + 
  geom_point() + 
  geom_hline(yintercept=-log10(0.05), linetype="dashed", color = "red") +
  labs(x="Spearman ρ", y="-log10(p-value)", color="Significance") + 
  theme_bw() + 
  geom_text_repel(data = subset(corr_results_h2, Pvalue < 0.05), 
                  aes(label=Species), 
                  size=3, 
                  box.padding = unit(0.35, "lines"), 
                  max.overlaps = 10, 
                  color = "black",
                  fontface = "bold",
                  segment.color = "grey",
                  segment.alpha = 0.7) +
  ggtitle("") + # log-H2 levels
  theme_pubclean() +
  theme(legend.position = "none")
p_h2_corr

# # Correct 'Species' names if needed
# corr_results_h2[154, "Species"] <- "Ruminococcaceae"

```
Correlation log plot -- **CH4**
```{r}
# Create a new column in the dataframe to indicate significance
corr_results_ch4$Significant <- ifelse(corr_results_ch4$Pvalue < 0.05, "Significant", "Not significant")

# Create the volcano plot
p_ch4_corr <- ggplot(corr_results_ch4, aes(x=Correlation, y=-log10(Pvalue), color=Significant)) + 
  geom_point() + 
  geom_hline(yintercept=-log10(0.05), linetype="dashed", color = "red") +
  labs(x="Spearman ρ", y="-log10(p-value)", color="Significance") + 
  theme_bw() + 
  geom_text_repel(data = subset(corr_results_ch4, Pvalue < 0.05), 
                  aes(label=Species), 
                  size=3, 
                  box.padding = unit(0.35, "lines"), 
                  max.overlaps = 10, 
                  color = "black",
                  fontface = "bold",
                  segment.color = "grey",
                  segment.alpha = 0.7) +
  ggtitle("") + # log-CH4 levels
  theme_pubclean() +
  theme(legend.position = "none")
p_ch4_corr

# # Correct 'Species' names if needed
# corr_results_ch4[154, "Species"] <- "Ruminococcaceae"

```

Combine all 3 log10 correlation plots into one figure
```{r}
# Vertical
p_h2ch4_corr <- egg::ggarrange(p_h2_corr, p_ch4_corr, p_ch4_to_h2_corr, labels = c("A", "B", "C"), 
                         ncol = 1, nrow = 3, heights = c(1,1,1))
p_h2ch4_corr

# Horizontal
# p_h2ch4_corr <- egg::ggarrange(p_h2_corr, p_ch4_corr, p_ch4_to_h2_corr, 
#                                labels = c("A", "B", "C"),
#                                ncol = 3, nrow = 1, 
#                                widths = c(1,1,1))
# p_h2ch4_corr

```
Save plot
```{r}
ggsave(filename = "~/Documents/Programming/Microbiome/microbiome-ocp/data-analyses/results/h2_ch4_corr/p_h2_ch4_corr-microab.png", plot = p_h2ch4_corr, width = 6, height = 12, dpi = 300)

```
